crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -365,6 +408,56 @@ impl<'db> TypeInferenceBuilder<'db> {
         self.types.definitions.insert(definition, function_ty);
     }
 
+    fn infer_parameters(&mut self, parameters: &ast::Parameters) {
+        let ast::Parameters {
+            range: _,
+            posonlyargs,
+            args,
+            vararg,
+            kwonlyargs,
+            kwarg,
+        } = parameters;
+
+        for param_with_default in posonlyargs {
+            self.infer_parameter_with_default(param_with_default);
+        }
+        for param_with_default in args {
+            self.infer_parameter_with_default(param_with_default);
+        }
+        if let Some(vararg) = vararg {
+            self.infer_parameter(vararg);
+        }
+        for param_with_default in kwonlyargs {
+            self.infer_parameter_with_default(param_with_default);
+        }
+        if let Some(kwarg) = kwarg {
+            self.infer_parameter(kwarg);
+        }
+    }
+
+    fn infer_parameter_with_default(&mut self, parameter_with_default: &ast::ParameterWithDefault) {
+        let ast::ParameterWithDefault {
+            range: _,
+            parameter,
+            default,
+        } = parameter_with_default;
+        self.infer_parameter(parameter);
+        if let Some(default) = default {
+            self.infer_expression(default);
+        }
```

Comment:
Nit: Maybe introduce a `infer_optional_expression(default.as_deref())`  method to reduce the `if let Some`

Reply:
Yeah I thought about this, I'll add it.